src/sumcheck/constraints.rs

```diff
@@ -0,0 +1,410 @@
+///                                    0  otherwise
+///                for 0 <= k < 3
+///
+/// Discussed in Section 6.6 [1].
+///
+/// # Bugs
+///
+/// This won't work in fields like GF(2^128) where P2 isn't literally two.
+///
+/// [1]: https://datatracker.ietf.org/doc/html/draft-google-cfrg-libzk-01#section-6.6
+fn lag_i<FE: CodecFieldElement>(i: FE, x: FE) -> FE {
+    // only lag_0, _1, _2 are defined
+    assert!(i == FE::ZERO || i == FE::ONE || i == FE::SUMCHECK_P2);
+
+    if x == i { FE::ONE } else { FE::ZERO }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        circuit::{Evaluation, tests::CircuitTestVector},
+        fields::fieldp128::FieldP128,
+    };
+
+    #[test]
+    fn longfellow_rfc_1_87474f308020535e57a778a82394a14106f8be5b() {
+        let (_, circuit) =
+            CircuitTestVector::decode("longfellow-rfc-1-87474f308020535e57a778a82394a14106f8be5b");
+
+        let witness_layout = WitnessLayout::new(
+            circuit.num_private_inputs(),
+            circuit.layers.iter().map(CircuitLayer::logw).collect(),
+        );
+
+        // This circuit verifies that 2n = (s-2)m^2 - (s - 4)*m. For example, C(45, 5, 6) = 0.
+        let evaluation: Evaluation<FieldP128> = circuit.evaluate(&[45, 5, 6]).unwrap();
+
+        // Matches session used in longfellow-zk/lib/zk/zk_test.cc
+        let mut proof_transcript = Transcript::new(b"test").unwrap();
+
+        let proof = Proof::new(
+            &circuit,
+            &evaluation,
+            &mut proof_transcript,
+            // Here we do _not_ fix the pad to zeroes in order to exercise symbolic manipulation.
+            FieldP128::sample,
+        )
+        .unwrap();
+
+        let mut constraint_transcript = Transcript::new(b"test").unwrap();
+        let constraints = ProofConstraints::from_proof(
+            &circuit,
+            evaluation.public_inputs(circuit.num_public_inputs()),
+            &mut constraint_transcript,
+            &proof,
+        )
+        .unwrap();
+
+        assert_eq!(
+            constraints.linear_constraint_lhs.len(),
+            3 * circuit.num_layers()
+                + circuit.logw_sum() * 2 * 2
+                + circuit.num_private_inputs()
+                + 2
+        );
+
+        for lhs_term in constraints.linear_constraint_lhs {
+            // All LHS terms should refer to an element of the RHS vector.
+            assert!(lhs_term.constraint_number < constraints.linear_constraint_rhs.len());
+            assert!(lhs_term.witness_index < witness_layout.length());
+        }
+
+        assert_eq!(
+            constraints.linear_constraint_rhs.len(),
+            circuit.num_layers() + 1
+        );
+
+        assert_eq!(
+            constraints.quadratic_constraints.len(),
+            circuit.num_layers()
+        );
+
+        // Transcripts should have received the same sequence of writes.
+        assert!(proof_transcript.compare_state(&constraint_transcript));
```

Comment:
This assertion is failing, so I must be missing a write to the transcript somewhere in the constraint generator. Or perhaps I'm wrong to expect the two transcripts to be in the same state.

Reply:
For now I refactored the sumcheck prover so we can toggle whether it runs in zk_test.cc compat mode, enabling both tests to pass.